paper round 3 (#183)
Skn0tt/datapact: paper/paper.md

    orcid: 0000-0003-4589-9809
    affiliation: 1
affiliations:
  - name: Hasso-Plattner-Institute for Digital Engineering, Digital Engineering Faculty, University of Potsdam, Potsdam, Germany
    index: 1
date: 30 June 2022
bibliography: paper.bib
---

# Summary

It is crucial to have a good understanding of the datasets that are being worked with. For static datasets, users can explore the dataset once, and document their findings for other team members to consume.
For frequently changing datasets, this may not be sufficient.
After exploring, analysing and documenting it once, an understanding needs to be continously verified to be up-to-date, and adjusted when it is not.
To address these problems, we developed `datapact`, a Python library that provides an ergonomic API for running assertions on DataFrames.

`datapact` is available under MIT license at https://github.com/skn0tt/datapact.

# Statement of need

When working with continually changing datasets, e.g. occupancy statistics for intensive care units during a pandemic, dataset shift (@moreno2012unifying) will occur frequently.

> Sometimes because of human errors in data entry, sometimes because of bugs, errors or outages in the data delivery, and sometimes because there are changes in the underlying data.

Reviewer: I would suggest rewriting this sentence to avoid so many "sometimes". For example:
```suggestion
Some common causes of data shift are human errors in data entry, bugs, errors or outages in the data delivery, or simply changes in the underlying data.
```

Author: love it!

If these changes stay unnoticed, they can make their way into subtle bugs, faulty visualisations and ultimately lead to suboptimal decisions.
The faster they get noticed, the faster analysis scripts, dashboards and data pipelines can be updated.

Without dedicated systems to detect them, changes can often go unnoticed for months.
`datapact` enables such a system to be built easily, so changes are noticed as the data comes in.
It is a Python library for expressing and verifying characteristics on DataFrames, and can be embedded right within the Python scripts and Jupyter notebooks that are already being used for analysis.

> Different tests are needed, both simple sanity checks ("the patient's age should be between 0 and 125") and more complex characteristics ("the age should be normally distributed").

Reviewer: ```suggestion
Different tests are supported, both simple sanity checks ("the patient's age should be between 0 and 125") and more complex characteristics ("the age should be normally distributed").
```

More specialized tests can easily be added via custom code.

> Existing projects addressing this need (e.g. Great Expectations (@Gong_Great_Expectations)) come with comparatively high entrance hurdles that are hard to overcome, particularly for non engineering heavy organisations.

Reviewer: ```suggestion
Existing projects addressing this need, e.g., Great Expectations (@Gong_Great_Expectations) come with comparatively high entrance hurdles that are hard to overcome, particularly for non engineering heavy organisations.
```

Author: ```suggestion
Existing projects addressing this need, e.g. Great Expectations (@Gong_Great_Expectations), come with comparatively high entrance hurdles that are hard to overcome, particularly for non engineering heavy organisations.
```

this looks better to me, and according to https://linguaholic.com/linguablog/comma-after-e-g/ it's okay to omit the comma after `e.g.`

Reviewer: Sounds good

With `datapact`, we provide an easily usable library that is designed to be used by researchers, industry data scientists and data engineers alike.
For teams there is `Datapact Track`: the optional web-app that tracks test results, and notifies about failures via E-Mail, Slack, Microsoft Teams, or PagerDuty.

# Method

Qualitative interviews across a variety of research and industry organisations were conducted to learn about the problem of unnoticed dataset shift.
After evaluating prior art, the idea of an assertion library was identified as a potential solution to the problem.

> To ensure that `datapact`'s developer-facing API is intuitive + ergonomic to use, the authors performed informal experiments where developers without prior knowledge about the problem were asked to explain what they think a given `datapact` snippet did.

Reviewer: I would avoid using `+` since it is not a mathematical expression.
```suggestion
To ensure that `datapact`'s developer-facing API is intuitive and ergonomic to use, the authors performed informal experiments where developers without prior knowledge about the problem were asked to explain what they think a given `datapact` snippet did.
```

To ensure high code quality, `datapact` maintains 100% code coverage via unit tests, automatically executed by our continuous integration.
Code style is linted via pylint (@pylint), pyright (@pyright) and Prettier (@prettier).
Documentation is auto-generated from docstrings (@goodger2010docstring).

# Example

> ![Statistical assertions on the Iris dataset.\label{fig:example}](screenshot.png)

Reviewer: In my last submission one of the reviewers asked me to create a minted environment for my code example. I would like to suggest a minted environment here as well, since it still displays nicely and is easier for readers to copy/paste rather than typing what is on the image. The results can still be images, but I would suggest adding a vector image, e.g., pdf or eps so it doesn't look pixelated on high resolution screens

Author: could you elaborate on what a minted image is?

this is a screenshot from VS Code, I don't see an easy way of making it a vector graphic without rebuilding it in a drawing program; which defeats the purpose of a screenshot. The image is already pretty high-resolution, but I could try upsampling it artificially if that helps.

Reviewer: About the minted environment, it automatically highlights code for you and it is still text that can be copy/pasted. For example [https://pt.overleaf.com/learn/latex/Code_Highlighting_with_minted](https://pt.overleaf.com/learn/latex/Code_Highlighting_with_minted).

The stuff after the code can be an image, but a screenshot might not produce the best quality. If you are using a jupyter notebook, you can open it on chrome and go to `Developer Tool > Rendering > Select 'screen' in Emulate CSS media`. Then you should be able to print a PDF with the screen CSS and it will be in vector format already. I guess you can also include the code in this vector image if the minted environment is too much effort.

Author: Oh, you mean a code block! Good to know that LaTeX calls that „minted“ :)

When using a codeblock, I think it wouldn‘t be obvious that this is from a Jupyter Notebook. I might be thinking to pragmatic here, a screenshot of the current resolution sounds like a fine solution to me.

Reviewer: I just checked and it was better only with the image since it is redundant now. Don't worry about it, unless the reviewers ask for the code block.

Author: okidoke 👍 reverted the change in https://github.com/Skn0tt/datapact/pull/183/commits/caf3fa066b03911b7d7a3cf837c4516b22788162.
